### PR TITLE
Remove custom metadata tool

### PR DESCRIPTION
**Pull request by ywelsch**
Adds a command-line tool to remove broken custom metadata from the cluster state. 

Relates to #48701

### CODE REVIEW

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/2

File: docs/reference/commands/node-tool.asciidoc [hunk 1]
```diff
@@ -20,16 +20,20 @@ bin/elasticsearch-node repurpose|unsafe-bootstrap|detach-cluster|override-versio
 [float]
 === Description
 
-This tool has five modes:
+This tool has six modes:
```
Comment (DaveCTurner) [inline docs/reference/commands/node-tool.asciidoc]: Maybe we should give up here :grin:

```suggestion
This tool has a number of modes:
```

File: docs/reference/commands/node-tool.asciidoc [hunk 2]
```diff
@@ -407,6 +429,33 @@ You can also use wildcards to remove multiple settings, for example using
...
+[float]
+==== Removing custom metadata from the cluster state
+
+If the on-disk cluster state contains custom metadata that prevents the node
+from starting up and loading the cluster state, you can run the following
+commands to remove this custom metadata.
+
+[source,txt]
+----
+node$ ./bin/elasticsearch-node remove-customs snapshot_lifecycle
```
Comment (DaveCTurner) [inline docs/reference/commands/node-tool.asciidoc]: Do users have a good way of knowing the name of the custom that should be removed? I.e. `snapshot_lifecycle` here.

File: server/src/main/java/org/elasticsearch/cluster/coordination/RemoveCustomsCommand.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        terminal.println(Terminal.Verbosity.VERBOSE, "Loading cluster state");
+        final Tuple<Long, ClusterState> termAndClusterState = loadTermAndClusterState(persistedClusterStateService, env);
+        final ClusterState oldClusterState = termAndClusterState.v2();
+        terminal.println(Terminal.Verbosity.VERBOSE, "custom metadata names: " + oldClusterState.metaData().customs().keys());
+        final MetaData.Builder metaDataBuilder = MetaData.builder(oldClusterState.metaData());
+        for (String customToRemove : customsToRemove) {
+            boolean matched = false;
+            for (ObjectCursor<String> customKeyCur : oldClusterState.metaData().customs().keys()) {
+                final String customKey = customKeyCur.value;
+                if (Regex.simpleMatch(customToRemove, customKey)) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/RemoveCustomsCommand.java]: Should we document that this uses wildcards, e.g. that `*` will remove all customs?

**Review by DaveCTurner — APPROVED** [review]
LGTM, some comments/questions on UX & docs but nothing major.

File: docs/reference/commands/node-tool.asciidoc (hunk 2 above)
Comment (ywelsch) [reply docs/reference/commands/node-tool.asciidoc]: This API is expert-level, not to be used to remove random stuff. I think the name here will be provided by us for specific support cases where we have previously seen a parse exception. Note that ES still has another way to remove the metadata, namely uninstalling the plugin that defines the corresponding metadata (we just drop the metadata at load time then).

File: server/src/main/java/org/elasticsearch/cluster/coordination/RemoveCustomsCommand.java (hunk 3 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/coordination/RemoveCustomsCommand.java]: This is documented, with the docs saying
> The tool takes a list of custom metadata names as parameters that should be removed, and also supports wildcard patterns.
I don't expect this to be used much, it's more to drop everything as last resort.

**Comment by raf64flo** [comment]
Hi, only a minor remark related to the title of this PR as well as in #50694 both of which appear in blog notes (https://www.elastic.co/blog/this-week-in-elasticsearch-and-apache-lucene-2020-01-17), I find them ambiguous and would find a naming like below better: 

> Tool to remove custom metadata

Anyway, thanks for that!

**Pull request merged**